### ISSUE TITLE
duplicate SFN entries lead to problems

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#2282]
When copying multiple files in DN,
not all of them are actually copied.

**Comment by stsp** [comment]
It turns out the problem is caused
by the duplicate SFN entries, eg.
`readme.txt` and `README.TXT`.
The program takes either one by
FindNext, then calls open()... and
opens another one.
I don't see any simple solution.
Maybe add some blacklist to mask
out such entries, and then `find_file()`
must be cluttered to skip them...

**Comment by andrewbird** [comment]
Is it possible to treat them like lfn, the first found gets returned as is and added to the cache, the second found after seeing there's a cache hit, gets tilded and that's returned instead?

**Comment by stsp** [comment]
This is not enough.
Besides avoiding dupes in FindNext,
you also need to make sure open()
(and everything else, like rename() etc)
also take the correct one.
Do you want to work on this?
Because I certainly do not. :)

**Comment by stsp** [comment]
I am planning to hide either one.
But even this must be consistently
done. I.e. the one you hide from
FindNext, should be the same you
hide from open().

**Comment by andrewbird** [comment]
> Do you want to work on this?

I'm not sure I have the time.

As I understand it there are two distinct scenarios:

1. FindFirst/Next (the directory is scanned and each result is returned in the order found)
2. Open/Creat/Rename/Delete  (I presume the file name is tried as is, then falls back if not found to a case insensitive directory search)

Almost the same problem happens with LFN names. What about say 'LongFileName.Txt' and 'lONGfILEnAME.tXT', I've set their content to the respective filename.
~~~
C:\>ver /r
comcom64 v0.3, dj64 (C) stsp
 Source ID: Non-git_build
 
Reported DOS version (Int21.3000): 5.00 OEM: FDh
Reported true DOS version (Int21.3306): 7.10
Version string (Int21.33FF): FDPP kernel 1.7 [GIT: ] (compiled Aug  7 2024)
C:\>type lONGfILEnAME.tXT
lONGfILEnAME.tXT
C:\>dir *.txt
 Volume in drive c is /home/ajb/dosemu/drive_c
 Directory of c:\
 
2024-10-03 22:01            17 LongFileName.Txt
2024-10-03 22:01            17 lONGfILEnAME.tXT
         2 file(s)             34 bytes
         0 dir(s)            19.7 GB free
C:\>type LONGFILENAME.TXT
LongFileName.Txt
~~~

So `type lONGfILEnAME.tXT` and `type LONGFILENAME.TXT` return different data. I'm not sure if the directory order is significant to which file data is actually returned when the name only matches insensitively. In an ideal world I think the first access be that a directory scan, or a file operation by name (Create, Open, Rename, Delete), should set the precedent for the insensitive match.

**Comment by stsp** [comment]
No, with SFN its much worse.
The program loses the track between
FindNext and open().
FindNext returns:
```
---all versions, documented fields---
 15h    BYTE    attribute of file found
 16h    WORD    file time (see #01665 at AX=5700h)
 18h    WORD    file date (see #01666 at AX=5700h)
 1Ah    DWORD   file size
```

So if program takes the `file size`
from here, then does open() and
read(), it suddenly finds it was fooled
about file size.

**Comment by stsp** [comment]
I've found the way to fix that w/o
hacks, provided the DOS app doesn't
mangle the case on its own.
If it does, the fix in #2283 is insufficient.
